tests/test_server.py

```diff
@@ -52,140 +71,152 @@ def test_add_device(client, name='steelseries-rival310'):
-    data = json.loads(response.data)
-    assert data == DeviceHandler.devices[id].hprof.dpi[dpi_id if
-        dpi_id != 'active' else DeviceHandler.devices[id].hprof.active_dpi]
+    data = response.json()
+    assert 'id' in data
+    assert 'shortname' in data
+    assert 'name' in data
+    assert 'input_nodes' in data
 
+    client.delete('/devices/{}'.format(id))
 
-def test_set_dpi(client, dpi_id=0):
+
+def test_dpi(client, dpi_id=0):
     from ratbag_emu.device_handler import DeviceHandler
 
-    id = test_add_device(client)
+    id = add_device(client)
 
-    DeviceHandler.devices[id].hprof.dpi[dpi_id if dpi_id != 'active' else
-        DeviceHandler.devices[id].hprof.active_dpi] = 6666
+    data = 6666
+    response = client.put('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id),
+                          json=data)
 
     response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id))
     assert response.status_code == 200
-    data = json.loads(response.data)
-    assert data == 6666
-
+    assert response.json() == data
 
-def test_get_active_dpi(client):
-    test_get_dpi(client, 'active')
+    client.delete('/devices/{}'.format(id))
 
 
-def test_set_active_dpi(client):
-    test_set_dpi(client, 'active')
+def test_active_dpi(client):
+    test_dpi(client, 'active')
 
 
-def test_get_led(client):
-    id = test_add_device(client)
-
-    response = client.get('/devices/{}/phys_props/leds/0'.format(id))
-    assert response.status_code == 200
-
-
-def test_set_led(client):
-    id = test_add_device(client)
+def test_led(client):
+    id = add_device(client)
 
     data = [
         0xFF,
         0xFF,
         0xFF
     ]
-    response = client.put('/devices/{}/phys_props/leds/0'.format(id),
-                          data=json.dumps(data),
-                          content_type='application/json')
+    response = client.put('/devices/{}/phys_props/leds/0'.format(id), json=data)
+    assert response.status_code == 200
+
+    response = client.get('/devices/{}/phys_props/leds/0'.format(id))
     assert response.status_code == 200
+    assert response.json() == data
+
+    client.delete('/devices/{}'.format(id))
 
 
-def test_device_event_raw_xy(client):
+def test_device_event(client):
     import os
 
-    from time import sleep
+    from time import time, sleep
     from threading import Thread
 
     import libevdev
     import fcntl
     import warnings
 
     from ratbag_emu.device_handler import DeviceHandler
+    from ratbag_emu.protocol.base import MouseData
 
-    id = test_add_device(client, 'generic-hidpp20')
-    sleep(0.1) # Give time for the event nodes to be created
+    id = add_device(client, 'generic-hidpp20')
+    input_nodes = []
 
-    DeviceHandler.devices[id].verbose = False
+    start = time()
+    while len(input_nodes) == 0:
+        input_nodes = client.get('/devices/{}'.format(id)).json()['input_nodes']
+        sleep(0.1)
+        if time() > start + 3:
+            break
```

Comment:
yay, polling!
Can't you make the `add_device` call blocking so that you have a full state before requesting the input nodes?

Reply:
How do I check if I have a full state? I can't reliably predict the amount of device nodes that will show up.

Reply:
You *can* know how many devices nodes you need by providing a full json description of your emulated device. You know then that you'll have a mouse, and/or a keyboard. And without having this json description, you already have a static definition of the report descriptors, so you already know before hand which capabilities will need to be present in the event nodes to have a fully ready device.

From the client point of view, you can't really get rid of polling unfortunately, unless your client waits for udev events.  But if the `add_device` is blocking to some extends, then it should be good enough to not hammer the server every 100ms to just get the new devices.

Reply:
/me checks the calendar

Oh, it's 2019. If every 100ms is "hammering" something, you need to upgrade your 286 to something newer, like a 386 ;)

but yeah, I agree, `add_device` should be smarter here.